[Filter] Performance increases.

mikeshardmind opened the pull request:
### Type

- [ ] Bugfix
- [X] Enhancement
- [ ] New feature

### Description of the changes

The filter was already using re to split words, this just does the entire search in re instead.

Commit:
[Filter] Performance increases.

The filter was already using re to split words, this just does the entire search in re instead.

A further improvement to this would cache patterns used and update them if the wordlist changes.

Tobotimus commented:
First issue with this is that there's a fatal change of behaviour - with an empty word list, it will return a set containing a single empty string, causing the message to be deleted. Obviously we can just exit early instead.

Secondly, I've timed these changes and they're not necessarily any faster than before, in fact in many cases it's significantly slower. They seem to be showing a performance increase when the word list is small and the message is large. However as the parameters go in the opposite direction (bigger word list / smaller message) it shows a bigger slowdown.

I tried to give it a fair benchmark by taking a random sample of messages and testing them against an auto-generated banned word list of increasing size. I got the random sample by using the last 10,000 messages in #support. Here's a graph of my results:
![graph](https://user-images.githubusercontent.com/7438501/53884678-446cb300-4070-11e9-8f28-c3f2bacabc91.png)

Tobotimus commented:
It also seems to be performing worse when the banned words and phrases themselves have a longer average length. In the previous test I used word lists in the form `["word1", "word2", ...]`, but here are my results using word lists in the form `["longerword1", "longerword2", ...]`:
![graph](https://user-images.githubusercontent.com/7438501/53885274-661a6a00-4071-11e9-885b-d3674881951e.png)

mikeshardmind commented:
1. The previous version was returning an empty set on no hits as well, this is *not* changing the API.
2. I can't reprofile this right now, but this was running *faster* with the test cases I used, so when I get home, I'll see if I can see what happened with that. 
3. This should show up as even faster on repeated use if the further changes which aren't made here to cache patterns are made, so I'll add those to this PR as well.

Commit:
Add a pattern cache

Commit:
exit on no-words

Commit:
formatting pass

mikeshardmind commented:
The majority of the time difference without the cache was the generator expression which created the pattern.

Adding a pattern cache makes the regex solution multiple orders of magnitude faster on any message other than the first received in a channel with a filter.

(Side by side comparison from discord conversation, running the same above tests on the updated version, but in a profiler instead of just a straight time count.)

![image](https://user-images.githubusercontent.com/22035876/53911533-ae5f6b00-4024-11e9-8510-ba19f970a742.png)

Context: `old_filter` is the behavior of the existing, the green block with re.compile at the same level is the initial message cost, and `new_filter` (again at the same level) is the amount of time excluding generating the cached pattern (not visible at that scale, extra image below):

![image](https://user-images.githubusercontent.com/22035876/53911627-fbdbd800-4024-11e9-8755-c90ce8c064d4.png)

Tobotimus commented:
RE: 1, You seem to have fixed the issue now, the issue was not returning an empty set on no hits. I said the issue was it returned a non-empty set on everything when filter was loaded but there were no banned words.

mikeshardmind commented:
yeah, I realized I misunderstood your meaning on that, and added that, but forgot to comment :sweat:

Commit:
keep the return type consistent, even though this doesnt break core since this is available to cogs

Commit:
...

Tobotimus commented on redbot/cogs/filter/filter.py:
```diff
@@ -279,6 +283,10 @@ def __unload(self):
 
+    def invalidate_cache(self, guild: discord.Guild, channel: discord.TextChannel = None):
+        """ Invalidate a cached pattern"""
+        self.pattern_cache.pop((guild, None), None)
```
Needs support for invalidating the cache in a channel as well

mikeshardmind replied:
Toby, the key is a tuple there. That's intended.

Tobotimus replied:
Why bother having a channel parameter if it's unused?

mikeshardmind replied:
Fixed....

Tobotimus commented:
Very nice, you managed to take it from linear time to constant time :grin: 
![graph](https://user-images.githubusercontent.com/7438501/53914918-d7c6d800-40b2-11e9-86cd-14618d076412.png)

Commit:
Quit being an idiot

Commit:
Slight further improvements, a fix, and restructure

Moved actual set creation out of the inner portion.
Reduced config lookups in case of no filter.
Fixed channel wordlist fetching.

Commit:
I really should go back to using a pre-commit hook for the style stuff
